### ISSUE TITLE
`Critical dependency: the request of a dependency is an expression` when using App dir.

### ISSUE THREAD

**Issue by guillermo-avalos** [aralroca/next-translate#1029]
Using Next-Translate v2.0.5 with Next.js v13.3.
Node v19.9, npm v9.6.4

While trying to setup Next-Translate following the instructions in this post: https://dev.to/aralroca/i18n-translations-in-nextjs-13s-app-dir-for-serverclient-components-4ek8 and copying exactly the code form the CodeSanbox example, I'm getting this error in the server console:

`Critical dependency: the request of a dependency is an expression` 

I've been using Next-Translate for several projects but this is the first time with the new App router.

My `next.config.js` file is as follows:

```
const nextTranslate = require('next-translate-plugin')
module.exports = nextTranslate({
experimental: { appDir: true },
})
```

`middleware.js`:
```
import { NextResponse } from 'next/server'
import i18n from './i18n'

export function middleware(request) {
  const locale = request.nextUrl.locale || i18n.defaultLocale
  request.nextUrl.searchParams.set('lang', locale)
  request.nextUrl.href = request.nextUrl.href.replace(`/${locale}`, '')
  return NextResponse.rewrite(request.nextUrl)
}
```

`I18n.js`:
```
module.exports = {
  locales: ['es', 'en'],
  defaultLocale: 'es',
  pages: {
    '*': ['common'],
  },
}
```

While trying to navigate to the main page I just get a `404` error. `/es` or `/en` doesn't change a thing.
Any ideas?

**Comment by guillermo-avalos** [comment]
The 404 error went away after placing both `layout.jsx` and `page.jsx` inside a `[locale]` folder after reading Next's docs.
It kinda works, but Node seems to leak memory and spawning several instances. The server console still shows the error.

**Comment by vdjurdjevic** [comment]
I am having the same issue.

**Comment by kasuta96** [comment]
> Critical dependency: the request of a dependency is an expression

Same!
The 404 error is occurring because Next.js 13.3 has made some changes to routing (13.2 works perfectly fine).
like @guillermo-avalos said. After move to inside [locale] folder, it does work, but it doesn't seem to be stable.

Btw, to use next-translate 2.0.5, I believe we should stick with Next.js 13.2 for now.

**Comment by nirtamir2** [comment]
Same problem here. next 13.4 & next-translate 2.0.5.
The i18n loads only on the client, and not on the server.

Also the code to sync date and lang cause runtime error:

```
Unhandled Runtime Error
TypeError: Cannot read properties of undefined (reading 'length')
```

```tsx
"use client";

import { useEffect } from "react";
import dayjs from "dayjs";
import "dayjs/locale/en";
import timezone from "dayjs/plugin/timezone";
import utc from "dayjs/plugin/utc";
import useTranslation from "next-translate/useTranslation";

interface Props {
  children: JSX.Element;
}

dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.tz.guess();

export const DateProvider = ({ children }: Props) => {
  const { lang } = useTranslation();

  useEffect(() => {
    dayjs.locale(lang);
  }, [lang]);

  return children;
};
```

Is there a way to HACK it somehow (even losing ssr) until we have a fix?

**Comment by alvesvin** [comment]
This is a [warning from webpack](https://webpack.js.org/guides/dependency-management/#require-with-expression). It seems webpack doesn't like imports with unpredictable paths at build time used by the [default loader in next-translate-plugin](https://github.com/aralroca/next-translate-plugin/blob/cd98cce0d32a7547c1daafeddbb9b5e2643e7e9f/src/utils.ts#L7). I don't know if it's really the case but it seems omitting [overwriteLoadLocales](https://github.com/aralroca/next-translate-plugin/blob/cd98cce0d32a7547c1daafeddbb9b5e2643e7e9f/src/utils.ts#L23) still works.

**Comment by alvesvin** [comment]
Related https://github.com/aralroca/next-translate-plugin/issues/9

**Comment by joaopedrodcf** [comment]
I'm also facing the same issue I'm currently using "next": "13.4.7" and "next-translate": "2.4.4" don't have a codesanbox prepared yet but here are the things I did:

This is what appears on my terminal:
<img width="747" alt="image" src="https://github.com/aralroca/next-translate/assets/11476152/3dc9ba4f-98e1-4b07-a06d-c9a552120e05">

And in the browser:
<img width="1510" alt="image" src="https://github.com/aralroca/next-translate/assets/11476152/3ef05089-e2a3-49f4-a839-53c0cf688c2a">


I tried to follow the docs and this article: https://dev.to/aralroca/i18n-translations-in-nextjs-13s-app-dir-for-serverclient-components-4ek8

middleware.ts
```ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';
import i18n from './i18n';

// /es/page-name -> rewrites to -> /es/page-name?lang=es
export function middleware(request: NextRequest) {
    const locale = request.nextUrl.locale || i18n.defaultLocale;
    request.nextUrl.searchParams.set('lang', locale);
    return NextResponse.rewrite(request.nextUrl);
}
```

i18n.js
```js
module.exports = {
    locales: ['en', 'pt', 'fr'],
    defaultLocale: 'en',
    pages: {
        '*': ['common'],
    },
};
```

next.config.mjs
```mjs
// @ts-check
import bundleAnalyzer from '@next/bundle-analyzer';
import nextTranslate from 'next-translate-plugin';

const analyzeEnabled = process.env.ANALYZE === 'true';

/** @type {import("next").NextConfig} */
const config = {
    reactStrictMode: true,
    swcMinify: true,
    webpack: (config, { isServer, webpack }) => {
        return config;
    },
};

const withBundleAnalyzer = bundleAnalyzer({
    enabled: analyzeEnabled,
});

export default withBundleAnalyzer(nextTranslate(config));
```

My app dir is currently like this:
<img width="352" alt="image" src="https://github.com/aralroca/next-translate/assets/11476152/2062639f-df52-43c1-88b9-7384269238d2">

**Comment by aralroca** [comment]
The 404 probably is for the middleware, I need to update this article, but after next.js 13.2 the middleware with i18n has an issue reported in Next.js but we updated using dynamic router [lang]. Probably you can remove the middleware, is not needed now to add the lang param and rewrite of the middleware is causing a 404 for this Next.js bug

**Comment by zakidze** [comment]
any solotion of this?????

**Comment by rolandjlevy** [comment]
@aralroca is there any solution for this problem? Can you help?

When I add loadLocaleFrom for the fallback, I get the same error: `Critical dependency: the request of a dependency is an expression` and can't find any solution for it. Is it a bug?

I am using next v12.2.5 and next-translate v2.4.4 and next-translate-plugin v2.4.4

This is what I added to the config. When I remove it, the error dissapears:

```js
loadLocaleFrom: async (lang, ns) => {
    const defaultLocale = 'en-gb';
    const defaultNs = await import(`./locales/${defaultLocale}/${ns}.json`);
    if (lang === defaultLocale) return defaultNs;
    const localNs = await import(`./locales/${lang}/${ns}.json`);
    return { ...defaultNs, ...localNs };
}
```

I get the same error when adding this version too:

```js
loadLocaleFrom: async (lang, ns) => {
    const defaultLocale = 'en-gb';
    const defaultNs = await import(
      `./locales/${defaultLocale}/${ns}.json`
    ).then((x) => x.default);
    if (lang === defaultLocale) return defaultNs;
    try {
      const localeNs = await import(`./locales/${lang}/${ns}.json`).then(
        (x) => x.default
      );
      return { ...defaultNs, ...localeNs };
    } catch (e) {
      return defaultNs;
    }
}
```

**Comment by aralroca** [comment]
@rolandjlevy upgrading to the latest version of `next-translate-plugin` should fix the `Critical dependency: the request of a dependency is an expression` from  a custom `loadLocaleFrom`

**Comment by LagoonProject** [comment]
I have the same problem, using "next" v13.4.9, and "next-translate" v2.5.2.

```
- wait compiling...
- warn ./node_modules/next/dist/pages/_app.js
Critical dependency: the request of a dependency is an expression
```

I don't have the 404 error though, since I moved the middleware.ts to the 'src' folder on the same level as 'app' folder.

**Comment by b-vetter** [comment]
I have the same problem. All pages inside the [lang] folder return a 404 and the warning "Critical dependency: the request of a dependency is an expression.
Is there already a solution how to fix it?

next 13.4.16
next-translate 2.5.3
next-translate-plugin 2.5.3

**Comment by roland-cromwell** [comment]
@aralroca I found a solution.

In case this helps someone else having this problem. The code below works for me and I am no longer getting the error: `Critical dependency: the request of a dependency is an expression`

I am using next v12.2.5 and next-translate v2.5.2 and next-translate-plugin v2.5.2

in i18n.js:

```js
 loadLocaleFrom: (lang, ns) => {
    const defaultLocale = 'en-gb';
    const defaultNs = require(`./locales/${defaultLocale}/${ns}.json`);
    if (lang === defaultLocale) return Promise.resolve(defaultNs);
    const localNs = require(`./locales/${lang}/${ns}.json`);
    return Promise.resolve({
      ...defaultNs,
      ...localNs,
    });
  },
```

**Comment by rolandjlevy** [comment]
> @aralroca is there any solution for this problem? Can you help?
> 
> When I add loadLocaleFrom for the fallback, I get the same error: `Critical dependency: the request of a dependency is an expression` and can't find any solution for it. Is it a bug?
> 
> I am using next v12.2.5 and next-translate v2.4.4 and next-translate-plugin v2.4.4
> 
> This is what I added to the config. When I remove it, the error dissapears:
> 
> ```js
> loadLocaleFrom: async (lang, ns) => {
>     const defaultLocale = 'en-gb';
>     const defaultNs = await import(`./locales/${defaultLocale}/${ns}.json`);
>     if (lang === defaultLocale) return defaultNs;
>     const localNs = await import(`./locales/${lang}/${ns}.json`);
>     return { ...defaultNs, ...localNs };
> }
> ```
> 
> I get the same error when adding this version too:
> 
> ```js
> loadLocaleFrom: async (lang, ns) => {
>     const defaultLocale = 'en-gb';
>     const defaultNs = await import(
>       `./locales/${defaultLocale}/${ns}.json`
>     ).then((x) => x.default);
>     if (lang === defaultLocale) return defaultNs;
>     try {
>       const localeNs = await import(`./locales/${lang}/${ns}.json`).then(
>         (x) => x.default
>       );
>       return { ...defaultNs, ...localeNs };
>     } catch (e) {
>       return defaultNs;
>     }
> }
> ```

@aralroca I found a solution 😀

In case this helps someone else having this problem. The code below works for me and I am no longer getting the error: `Critical dependency: the request of a dependency is an expression`

I am using next v12.2.5 and next-translate v2.5.2 and next-translate-plugin v2.5.2

in i18n.js:

```js
 loadLocaleFrom: (lang, ns) => {
    const defaultLocale = 'en-gb';
    const defaultNs = require(`./locales/${defaultLocale}/${ns}.json`);
    if (lang === defaultLocale) return Promise.resolve(defaultNs);
    const localNs = require(`./locales/${lang}/${ns}.json`);
    return Promise.resolve({
      ...defaultNs,
      ...localNs,
    });
  },
```

**Comment by mirik999** [comment]
> 

did you solve this problem ? i got the same error ((

**Comment by mirik999** [comment]
> I'm also facing the same issue I'm currently using "next": "13.4.7" and "next-translate": "2.4.4" don't have a codesanbox prepared yet but here are the things I did:
> 
> This is what appears on my terminal: <img alt="image" width="747" src="https://user-images.githubusercontent.com/11476152/249236884-3dc9ba4f-98e1-4b07-a06d-c9a552120e05.png">
> 
> And in the browser: <img alt="image" width="1510" src="https://user-images.githubusercontent.com/11476152/249238481-3ef05089-e2a3-49f4-a839-53c0cf688c2a.png">
> 
> I tried to follow the docs and this article: https://dev.to/aralroca/i18n-translations-in-nextjs-13s-app-dir-for-serverclient-components-4ek8
> 
> middleware.ts
> 
> ```ts
> import { NextResponse } from 'next/server';
> import type { NextRequest } from 'next/server';
> import i18n from './i18n';
> 
> // /es/page-name -> rewrites to -> /es/page-name?lang=es
> export function middleware(request: NextRequest) {
>     const locale = request.nextUrl.locale || i18n.defaultLocale;
>     request.nextUrl.searchParams.set('lang', locale);
>     return NextResponse.rewrite(request.nextUrl);
> }
> ```
> 
> i18n.js
> 
> ```js
> module.exports = {
>     locales: ['en', 'pt', 'fr'],
>     defaultLocale: 'en',
>     pages: {
>         '*': ['common'],
>     },
> };
> ```
> 
> next.config.mjs
> 
> ```js
> // @ts-check
> import bundleAnalyzer from '@next/bundle-analyzer';
> import nextTranslate from 'next-translate-plugin';
> 
> const analyzeEnabled = process.env.ANALYZE === 'true';
> 
> /** @type {import("next").NextConfig} */
> const config = {
>     reactStrictMode: true,
>     swcMinify: true,
>     webpack: (config, { isServer, webpack }) => {
>         return config;
>     },
> };
> 
> const withBundleAnalyzer = bundleAnalyzer({
>     enabled: analyzeEnabled,
> });
> 
> export default withBundleAnalyzer(nextTranslate(config));
> ```
> 
> My app dir is currently like this: <img alt="image" width="352" src="https://user-images.githubusercontent.com/11476152/249237937-2062639f-df52-43c1-88b9-7384269238d2.png">

did you solve this problem ? i got the same error ((

**Comment by csexplorer** [comment]
the same issue!!!!

**Comment by roland-cromwell** [comment]
> > @aralroca I found a solution 😀
> > In case this helps someone else having this problem. The code below works for me and I am no longer getting the error: `Critical dependency: the request of a dependency is an expression`
> 
> Tested and doesn't work in my environment. Still the same error/warning.

@b-vetter can you share the contents of your i18n.js and next.config,js files?

**Comment by b-vetter** [comment]
@roland-cromwell 

i18n.js
```javascript
module.exports = {
    locales: ["en", "de"],
    defaultLocale: "en",
    pages: {
      "*": ["common"],
      "/[lang]": ["home"]
    },
    loadLocaleFrom: (lang, ns) => {
      const defaultLocale = 'en';
      const defaultNs = require(`./locales/${defaultLocale}/${ns}.json`);
      if (lang === defaultLocale) return Promise.resolve(defaultNs);
      const localNs = require(`./locales/${lang}/${ns}.json`);
      return Promise.resolve({
        ...defaultNs,
        ...localNs,
      });
    }
  };
```

next.config.js
```javascript
/** @type {import('next').NextConfig} */

const nextTranslate = require('next-translate-plugin')

const dns = require("dns");
dns.setDefaultResultOrder("ipv4first")

const nextConfig = nextTranslate({
  images: {
    domains: ['avatars.githubusercontent.com', 'avatar.vercel.sh']
  },
  experimental: {
    appDir: true,
    serverComponentsExternalPackages: ['@tremor/react']
  },
  async headers() {
    return [
      {
        source: "/api/:path*",
        headers: [
          // some header configuration
        ],
      }
    ]
  },
});

module.exports = nextConfig;
```

**Comment by abelchee** [comment]
I got the same error in https://github.com/aralroca/next-translate/tree/master/examples/with-app-directory when using next 13.4.19

next 13.4.7 works fine in the example

![image](https://github.com/aralroca/next-translate/assets/13580085/a2f8ac07-2022-45de-b749-12c3e074f459)

**Comment by cmaerz** [comment]
Don't know whats happening there, but just do this:
![image](https://github.com/aralroca/next-translate/assets/1479186/562d3b88-c2e4-4dda-b7e3-6d00248a328c)
It works fine somehow?! Even with 13.4.19

**Comment by abelchee** [comment]
> Don't know whats happening there, but when I just do this: ![image](https://user-images.githubusercontent.com/1479186/263960943-562d3b88-c2e4-4dda-b7e3-6d00248a328c.png) It works fine somehow?! Even with 13.4.19

It doesn't work for me

**Comment by b-vetter** [comment]
> I got the same error in https://github.com/aralroca/next-translate/tree/master/examples/with-app-directory when using next 13.4.19
> 
> next 13.4.7 works fine in the example

Works also for me with 13.4.7. So is it a problem at next-js? When can we expect a fix on newer versions?

**Comment by bluelakee02** [comment]
I created a fix https://github.com/aralroca/next-translate-plugin/pull/72
Here is a patch if somebody needs it right away https://github.com/bluelakee02/next-translate-plugin-patch

**Comment by bluelakee02** [comment]
@aralroca OK, now this can be probably closed

**Comment by aralroca** [comment]
> @aralroca OK, now this can be probably closed

Sure 😊 people, now is already solved in 2.6.1 thanks to @bluelakee02

**Comment by AndryKK** [comment]
@bluelakee02 can u give access to https://github.com/bluelakee02/next-translate-plugin-patch ?
